array.h

```diff
@@ -2757,6 +2756,33 @@ void generate(array<T, Shape, Alloc>& dst, Generator&& g) {
   generate(dst.ref(), g);
 }
 
+/** Fill the `dst` array or array_ref with the result of calling a "pattern
+ * function" `fn`, which takes a single shape::index_type argument. The order in which `fn` is
+ * called is the same as `shape_traits<Shape>::for_each_index`. */
+template <typename T, typename Shape, class Fn>
+NDARRAY_HOST_DEVICE void transform_index(const array_ref<T, Shape>& dst, Fn&& fn) {
+  using index_type = typename Shape::index_type;
+  for_each_index(dst.shape(), [&, fn = std::move(fn)](index_type idx) { dst[idx] = fn(idx); });
```

Comment:
> Thinking about this a bit more... consider if you should just use `for_each_index` directly. Here's why:
> 
> * This is really a super thin layer around `for_each_index`, but it removes important functionality, like the ability to control the order of the loops `for_each_index<a, b, c>(...)`.
> * The existing similar helper `generate` exists because it can flatten the shape before traversing the loop, and `for_each_value` is maybe a bit more tricky to use directly.
> 
> If you still want to add it, I'm fine with it, but I personally would probably tend to just use `for_each_index` directly.

I get the argument. But `for_each_index` and `generate` with the lambdas and such really is quite a lot of typing.